File: client/python/gamium/gamium/utils/wait.py
import inspect
from datetime import datetime
import time
from gamium.protocol.generated.Types.ErrorCode import ErrorCode
from gamium.condition.condition import Condition
from gamium.condition.wait_condition import WaitCondition
from gamium.errors.gamium_error import GamiumError
from gamium.igamium_client import IGamiumClient
from gamium.options.wait_options import WaitOptions
from gamium.utils.generics import T
from gamium.utils.time import current_time_ms


def wait_generic(client: IGamiumClient, condition: WaitCondition[T], options: WaitOptions) -> T:
    if client.is_connected() == False:
        raise GamiumError(ErrorCode.Disconnected, "Not connected")

    def call_condition(condition: WaitCondition[T]):
        if callable(condition):
            signature = inspect.signature(condition)
            parameter_count = len(signature.parameters)
            if parameter_count > 0:
                return condition(client)  # type: ignore # no way to let mypy know that this has parameter
            else:
                return condition()  # type: ignore # no way to let mypy know that this hasn't  parameter
        else:
            return condition.func(client)

    start_time = current_time_ms()
    last_call_time = start_time
    is_timeouted = False

    for i in range(0, 99999):
        try:
            if current_time_ms() - start_time > options.timeout_ms:
                is_timeouted = True
            last_call_time = current_time_ms()
            condition_ret = call_condition(condition)
            if condition_ret:
                return condition_ret
        except Exception as e:
            if is_timeouted == True:
                raise e
            if options.ignore_error == False:
                raise e
        remain_interval_ms = options.interval_ms - (current_time_ms() - last_call_time)
        if 1 < remain_interval_ms:
            time.sleep(remain_interval_ms / 1000)

    raise GamiumError(ErrorCode.Timeout, "timeout")

File: client/typescript/gamium/src/common/internal/utils/wait.ts
import { GamiumWaitError } from '../../errors/gamium-wait-error';
import { GamiumClient } from '../../gamium-client';
import { DefaultWaitOptions, WaitOptions } from '../../options/wait-options';
import { ErrorCode } from '../../protocols/generated';
import { WaitCondition } from '../../wait-condition';
import { delay } from '../functions';

export const waitGeneric = async <Type>(
  gamiumClient: GamiumClient,
  condition: WaitCondition<Type>,
  option: Partial<WaitOptions> = DefaultWaitOptions(),
): Promise<Type> => {
  if (!gamiumClient.connected) {
    throw new GamiumWaitError(ErrorCode.Disconnected, 'notconnected');
  }
  const optionMixed: WaitOptions = { ...DefaultWaitOptions(), ...option };
  const callCondition: (cond: typeof condition) => Promise<Type> = async (condition) => {
    if ('then' in condition) {
      return await condition;
    }
    if ('message' in condition && 'contidionFunc' in condition) {
      return await condition.contidionFunc(gamiumClient);
    }
    const conditionRet = condition(gamiumClient);
    if ('then' in conditionRet) {
      return await conditionRet;
    }
    return conditionRet;
  };
  const startTime = Date.now();
  let lastCallTime = Date.now();

  let i = 0;
  const error = new GamiumWaitError(ErrorCode.Timeout, 'wait timeout');
  error.waitOptions = optionMixed;

  for (i = 0; i < Infinity; i++) {
    error.tryCount = i;
    try {
      if (Date.now() - startTime > optionMixed.timeoutMs) {
        break;
      }
      lastCallTime = Date.now();
      const conditionRet = await callCondition(condition);
      error.innerResult = conditionRet;
      if (conditionRet) {
        return conditionRet;
      }
    } catch (e: unknown) {
      error.cause = e as Error;
      if (!optionMixed.ignoreError) {
        break;
      }
    }
    const remainIntervalMs = optionMixed.intervalMs - (Date.now() - lastCallTime);
    if (1 < remainIntervalMs) {
      await delay(remainIntervalMs - 1);
    }
  }
  throw error;
};
